javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/FunctionBodyFeatures.qll

```diff
@@ -127,22 +127,29 @@ ASTNode getAnASTNodeWithAFeature(Function f) {
   result = getAnASTNodeToFeaturize(f)
 }
 
+int getNumCharsInFunction(Function f) {
+    result = strictsum(int i |
+      exists(ASTNode node | node = getAnASTNodeWithAFeature(f) and i = getTokenizedAstNode(node).length()) |
+      i
+    )
+}
+
+// Evaluator string limit is 5395415 characters. We choose a limit lower than this.
+private int getMaxChars() { result = 1000000 }
+
+Function getFeaturizableFunction(Function f) {
+   result = f and getNumCharsInFunction(f) <= getMaxChars()
+}
+
 /**
  * Returns a featurized representation of the function that can be used to populate the
  * `enclosingFunctionBody` feature for an endpoint.
  */
 string getBodyTokensFeature(Function function) {
-  // Performance optimization: If a function has more than 256 body subtokens, then featurize it as
-  // absent. This approximates the behavior of the classifer on non-generic body features where
-  // large body features are replaced by the absent token.
-  //
-  // We count nodes instead of tokens because tokens are often not unique.
-  strictcount(ASTNode node |
-    node = getAnASTNodeToFeaturize(function) and
-    exists(getTokenizedAstNode(node))
-  ) <= 256 and
-  result =
-    strictconcat(Location l, string token |
+  // Performance optimization: If a function has more than getMaxChars() characters in its body subtokens, 
+  // then featurize it as absent.
+  function = getFeaturizableFunction(function) and
+  result = strictconcat(Location l, string token |
```

Comment:
This changes the behaviour of the function body features potentially quite considerably.  The AST node limit was a factor that noticeably affected end-to-end performance, and using a different restriction here could have important implications on the runtime of the ML-powered queries.

We need to validate the runtime implications of this change before we merge it.  Furthermore, we should either (a) stick with the removal of the AST node limit, train a new model on the new features and run an end-to-end evaluation, or (b) reintrodue the AST node limit, so the feature values should not change considerably.

Reply:
- Will reinstate the AST node limit
- Will collect timing measures
- Will run end-to-end evaluation